cli/lsp/cache.rs

```diff
@@ -117,3 +123,97 @@ impl CacheServer {
     rx.await?
   }
 }
+
+/// Calculate a version for for a given path.
+pub(crate) fn calculate_fs_version(path: &Path) -> Option<String> {
+  let metadata = fs::metadata(path).ok()?;
+  if let Ok(modified) = metadata.modified() {
+    if let Ok(n) = modified.duration_since(SystemTime::UNIX_EPOCH) {
+      Some(n.as_millis().to_string())
+    } else {
+      Some("1".to_string())
+    }
+  } else {
+    Some("1".to_string())
+  }
+}
+
+/// Populate the metadata map based on the supplied headers
+fn parse_metadata(
+  headers: &HashMap<String, String>,
+) -> HashMap<MetadataKey, String> {
+  let mut metadata = HashMap::new();
+  if let Some(warning) = headers.get("x-deno-warning").cloned() {
+    metadata.insert(MetadataKey::Warning, warning);
+  }
+  metadata
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) enum MetadataKey {
+  /// Represent the `x-deno-warning` header associated with the document
+  Warning,
+}
+
+#[derive(Debug, Clone)]
+struct Metadata {
+  values: Arc<HashMap<MetadataKey, String>>,
+  version: Option<String>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct CacheMetadata {
+  cache: http_cache::HttpCache,
+  metadata: Arc<Mutex<HashMap<ModuleSpecifier, Metadata>>>,
```

Comment:
This change highlights that the cache being disconnected from the data in the `Arc<Mutex<_>>` is a little sketch for the sake of the snapshots, but this state being in sync is currently enforced by the "lsp mutex". In the future, if we allow multiple threads in, this state being in sync should be further enforced by requiring mutability on `set_location`, so we're good.

We may consider some improvements to this in the future to ensure the two are forced to stay in sync when a snapshot occurs so the code is slightly more resilient to changes around it, but this could be a very minor future improvement.